src/services/WaveformDataUtils.js

```diff
@@ -51,55 +51,47 @@ export default class WaveformDataUtils {
 
   /**
    * Add a new segment to Peaks when a new timespan is created
-   * @param {Array} smData - updated structured metadata
+   * @param {Object} newSpan - new span created for the user input
    * @param {Object} peaksInstance - peaks instance for the waveform
    */
-  insertNewSegment(smData, peaksInstance) {
-    const allSpans = structMetadataUtils.getItemsOfType('span', smData);
-    const segments = peaksInstance.segments.getSegments();
-    const newSpan = allSpans.filter(
-      span => !segments.some(segment => segment.id === span.id)
-    );
-    const { begin, end, label, id } = newSpan[0];
+  insertNewSegment(newSpan, peaksInstance) {
+    const { begin, end, label, id } = newSpan;
     peaksInstance.segments.add({
       startTime: structMetadataUtils.toMs(begin) / 1000,
       endTime: structMetadataUtils.toMs(end) / 1000,
       labelText: label,
       id: id
     });
 
-    return this.updateSegmentColors(peaksInstance, smData);
+    return peaksInstance;
   }
 
   /**
    * Delete the corresponding segment when a timespan is deleted
    * @param {String} id - ID of the segment that is being deleted
-   * @param {Array} smData - updated structured metadata
    * @param {Object} peaksInstance - peaks instance for the current waveform
    */
-  deleteSegment(id, smData, peaksInstance) {
+  deleteSegment(id, peaksInstance) {
     peaksInstance.segments.removeById(id);
-    return this.updateSegmentColors(peaksInstance, smData);
+    return peaksInstance;
   }
 
   /**
    * Update the colors of the segment to alternate between colors in Avalon color pallette
    * @param {Object} peaksInstance - current peaks instance for the waveform
-   * @param {Array} allItems - updated structured metadata
    */
-  updateSegmentColors(peaksInstance, allItems) {
-    const allSpans = structMetadataUtils.getItemsOfType('span', allItems);
-    const segments = peaksInstance.segments.getSegments();
-
+  rebuildPeaks(peaksInstance) {
+    let clonedSegments = peaksInstance.segments
+      .getSegments()
+      .sort((x, y) => x.startTime - y.startTime);
     peaksInstance.segments.removeAll();
-
-    // Update the colors of each segment
-    segments.forEach(segment => {
-      let id = segment.id;
-      let index = allSpans.map(span => span.id).indexOf(id);
+    let index = 0;
+    clonedSegments.forEach(segment => {
```

Comment:
Here's a little trick.  In the `forEach()` function, you get the `index` value for free as the 2nd parameter.  So you could lose the `index` variable declaration and incrementing, and do `clonedSegments.forEach((segment, index) => {...`

Reply:
Really!! I feel so dumb!

Reply:
Ha, JS hides all kinds of extra parameters in some of their core functions that are used a lot, or not much.  Your logic is perfect, just pointing out a tiny little trick.